python/lsst/validate/base/metric.py

```diff
@@ -1,21 +1,112 @@
 # See COPYRIGHT file at the top of the source tree.
 from __future__ import print_function, division
-from past.builtins import basestring
 
-import operator
+import os
+import glob
 from collections import OrderedDict
 import yaml
 
 from .jsonmixin import JsonSerializationMixin
-from .datum import Datum
-from .spec import Specification
 
+__all__ = ['Metric', 'MetricSet', 'MetricRepo', 'load_metrics']
 
-__all__ = ['Metric', 'load_metrics']
+
+class MetricRepo(object):
+    """A collection of MetricSets, each identified by name.
+
+    Parameters
+    ----------
+    path : `str`
+        path to the directory defining this MetricRepo
+    metric_sets : `dict`
+        dict of name: `MetricSet`
+    """
+    path = None
+    """`str` giving the path to the directory defining this MetricRepo."""
+
+    metric_sets = None
+    """`dict` of all the MetricSets defined in this repo, identified by name."""
+
+    def __init__(self, path, metric_sets):
+        self.path = path
+        self.metric_sets = metric_sets
+
+    @classmethod
+    def from_path(cls, path):
+        metric_sets = {}
+        for path in glob.glob(os.path.join(path, '*.yaml')):
+            name = os.path.splitext(os.path.basename(path))[0]
+            with open(path) as f:
+                metric_sets[name] = MetricSet.from_yaml(name, yaml.load(f))
+        return cls(path, metric_sets)
+
+    def __getitem__(self, key):
+        return self.metric_sets[key]
+
+    def __len__(self):
+        return len(self.metric_sets)
+
+    def __contains__(self, key):
+        return key in self.metric_sets
+
+    def __str__(self):
+        items = ",\n".join(str(self.metric_sets[k]) for k in sorted(self.metric_sets))
+        return "{0.path}: {{\n{1}}}".format(self, items)
+
+
+class MetricSet(object):
+    """A collection of metrics, each identified by name.
+
+    Parameters
+    ----------
+    name : `str`
+        The name of this metric set, usually the name of an LSST package
+    metrics : `dict` of `str`: `Metric`
+        A `dict` of names and their associated metrics.
+    """
+
+    name = None
+    """Name of the MetricSet, usually the name of a package."""
+
+    metrics = None
+    """{name: metric} dict of the metrics"""
+
+    def __init__(self, package_name, metrics):
+        self.name = package_name
+        self.metrics = metrics
+
+    @classmethod
+    def from_yaml(cls, name, yaml):
+        metrics = {}
+        for subname in yaml:
+            metric_name = "{}.{}".format(name, subname)
+            m = Metric.from_yaml(subname, yaml_doc=yaml)
+            metrics[metric_name] = m
+        return cls(name, metrics)
+
+    @classmethod
+    def from_list(cls, name, metric_list):
```

Comment:
I've changed it so `__init__` takes either a list or a dict: I think it needs a dict so that `from_yaml` can pass it a fully-constructed thing. If your new `Name` takes care of all that "automatically", that'd be great.